### DISCUSSION TITLE
Adding ChargePoint capture & overlay for PHEV/EV vehicles

### DISCUSSION THREAD

**Discussion by clukas1967** [jasonacox/Powerwall-Dashboard#242] [Ideas]
I have a non-Tesla PHEV and a ChargePoint home L2 charger.  A search of this forum for "PHEV" or "BEV" comes up empty, which seems odd since there must a lot of Tesla car owners here.  Right now we just have the one car but if California has its way that will probably be two by 2030.

On PW Dashboard, I can make out when I charge the car due to its characteristic 3.7 kW max current.  In my house there are basically 3 load signatures as follows:

<img width="167" alt="image" src="https://user-images.githubusercontent.com/47096023/231352239-b84eca91-ca60-4f01-8093-00727da0e738.png">

My base load is ~1.2 kW which is notionally steady.  When I charge the car, this jumps up to 1.2+3.7 = ~4.9 kW for however long it takes.  And we have a hot tub that stirs every 6 hours which is the brief high spike in the middle of the car charge shown in the figure.

Now even though I can separate everything by eyeball today, it would be nice/useful to have a separate widget showing just my car-specific charging activities.  And in a household with multiple PHEV/BEVs it would be even better to be able to show that consumption by car over time in different colors say.

The ChargePoint app does appear to have my history in it and they have an API.  There's another Git project that deals with it, though nothing I can find for visualization.   https://github.com/lefthandwriter/ChargePointAPI

Here's what ChargePoint shows for the same interval shown above in PW Dashboard.

![image](https://user-images.githubusercontent.com/47096023/231354223-9e3c963c-24a4-4e92-8a9e-b122d95969c7.png)
What I have in mind is a Grafana widget that looks like this, allowing multiple cars to be overlaid on the same timeline in different colors (and stacking where necessary for simultaneous charging).  

It would also be possible to import all the history just like with the Powerwall.  It seems that all of my ChargePoint history is available and clickable similar to Tesla app.

![image](https://user-images.githubusercontent.com/47096023/231354091-154672a5-b734-4c50-9793-296a381931e4.png)

Like the ChargePoint app, it would also be cool impute energy cost, or estimated distance from the total energy, with an eye towards have a single view of climate impact via PWD.

Has anybody messed around with something like this?   For those of you running PWD and have Tesla cars, are you doing any type of co-visualization?

Thanks!

**Comment by jasonacox** [comment]
I haven't see anyone explore this yet @clukas1967 .  Thanks for opening the discussion. 

I wonder if there is a way to get this data from the Tesla Owner API (via _teslapy_ library) similar to the Tesla-History script here https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools.

**Reply by mcbirse** [reply]
Hi @clukas1967 - just wondering if you ever made any progress getting ChargePoint data into InfluxDB / Powerwall Dashboard using the ChargePoint API?

I recently started working on adapting the Tesla-History script to get data from the myenergi API for Zappi chargers into InfluxDB. Once finished, it may be possible to alter the script further to get data from the ChargePoint API.

If you have any information or can help with the ChargePoint API specifics and/or testing, you might want to join the discussion here: #382

**Comment by youzer-name** [comment]
> Has anybody messed around with something like this? For those of you running PWD and have Tesla cars, are you doing any type of co-visualization?
> 
> Thanks!

I am doing something vaguely similar, but it won't be directly applicable to you as it relies on Teslamate which is a project for pulling data for Tesla cars.

Since Teslamate puts all the data for the cars in a PSQL database, I set up a data source in Grafana for that database and pull the charging status of each of my cars so I can display it on the main energy use graph. One of my cars is white and one is blue, so I display a blue line or white line depending which is charging (or pre-conditioning). Where you see light blue, it is the blue and white both charging at the same time.

The orange line you see is when my LG ThinQ clothes dryer is running, which I can pull from the logs of my home automation hub (Hubitat) that is connected to the LG API.

I don't need to bring cost/use stats into PW Dashboard because Teslamate already does that (and more) for the cars.

![image](https://user-images.githubusercontent.com/57046429/232259529-6673ad96-1bbf-42a9-ba87-62188c3cdb8f.png)

If anyone with Teslamate wants to do the same, I can share the query and field overrides that I added to the panel for this.

**Reply by jcomand** [reply]
It would be great to suck car charging data directly from the Tesla API.  https://www.teslaapi.io/

**Reply by jasonacox** [reply]
We could do something similar to what @mcbirse has done to get Solar Only data from Tesla API (see https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/solar-only ) create a container that polls this data and pushes it into InfluxDB as well.

**Reply by jcomand** [reply]
yes! or maybe more productively, since teslamate does this already, could merge this project into the same container as teslamate. (I do not know much about docker so I have no idea how hard this is.) The both use grafana, and when you are running both projects, it keeps asking for the password when you switch tabs.  But anyways, the databases are obviously in separate containers.

**Reply by youzer-name** [reply]
@jasonacox I don't think you can do something like the Solar Only data. Car history isn't available from the API. You either log it in real time, or you don't have access to it. People have asked for a way to fill in missing history in Teslamate, and it just isn't currently available from the API. 

@jcomand I don't think merging them into one stack really gets you anywhere. That wouldn't merge the two instances of Grafana. You'd just have two instances running on different ports in a single stack instead of in two different stacks. It would introduce all sorts of potential issues with updating this project whenever Teslamate updates.

What you seem to be asking to do is to merge all the dashboards (and datasources) into one instance of Grafana. You could already do that by exporting all 18 of the dashboards from Teslamate and importing them into your Powerwall Dashboard Grafana instance. 

The only added complication due to them being in different stacks is that you have allow the Grafana in the Powerwall Dashboard stack to connect to the database in the Teslamate stack. I did this by creating an overlay network and putting Grafana and the PostgreSQL database into that network. You could probably modify the Teslamate docker-compose to have that container expose a port and accomplish the same thing, but I haven't tried that.

I am able to have both Grafana instances open in different tabs and don't get prompted for my password when switching. Both of my Grafana instances are running on the same Raspberry Pi in Docker. The one for Teslamate is behind the Traefik proxy, so I access it by an external DDNS URL (https://myddnsurl:port) while I access the Powerwall Dashboard Grafana at it's local address (http://pi-lan-ip:port). Maybe that is why switching tabs works for me, but I would have expected two Grafana instances at the same IP but on different ports to act the same way.

**Comment by clukas1967** [comment]
Sorry have not had time to delve into it.
